plugins/copy-to-clipboard/prism-copy-to-clipboard.js

```diff
@@ -35,13 +35,39 @@
 		head.appendChild(script);
 	}
 
+	function getSettings(start, defaults) {
```

Comment:
From the method signature, I would have guessed that `defaults` is a readonly value. As it is right now, `getSettings` takes the `defaults` object, modifies it, and then returns it.

Either make this a pure function or don't return anything (and rename the function to better reflect its actual behaviour).

Also, please add a little JSDoc to explain the function. (Maybe even some types)

Reply:
We can simplify this further by moving the `defaults` up here and using them in scope. Agree with everything else.

Reply:
OK, I’ll make it pure in a backward-compatible manner.

Reply:
> We can simplify this further by moving the defaults up here and using them in scope.

I'm also for that. This makes the function even simpler.

```js
/**
 * Traverses up the DOM tree to find data attributes that override the default plugin settings.
 *
 * @param {Element} startElement An element to start from.
 * @returns {Settings} The plugin settings.
 *
 * @typedef {Record<"copy" | "copy-error" | "copy-success" | "copy-timeout", string | number>} Settings
 */
function getSettings(startElement) {
	/** @type {Settings} */
	var settings = {
		'copy': 'Copy',
		'copy-error': 'Press Ctrl+C to copy',
		'copy-success': 'Copied!',
		'copy-timeout': 5000
	};

	var prefix = 'data-prismjs-';
	for (var key in settings) {
		var attr = prefix + key;
		var element = startElement;
		while (element && !element.hasAttribute(attr)) {
			element = element.parentElement;
		}
		if (element) {
			settings[key] = element.getAttribute(attr);
		}
	}

	return settings;
}
```

Reply:
OK, I’ll follow the way.

Just one more point. `jsdoc` doesn’t parse this tag.

```javascript
/**
 * @typedef {Record<"copy" | "copy-error" | "copy-success" | "copy-timeout", string | number>} Settings
 */
```
`jsdoc` doesn’t parse this either.

```javascript
/**
 * @typedef {{"copy": String, "copy-error": String, "copy-success": String, "copy-timeout": Number}} Settings
 */
```

The only way to generate a list of properties is to use a complex type definition (the last example from [the table](https://jsdoc.app/tags-type.html#overview)):

```javascript
/**
 * @typedef Settings
 * @type {Object}
 * @property {String} copy A copy-to-clipboard message.
 * @property {String} copy-error A copying error message.
 * @property {String} copy-success A message for a successful copying.
 * @property {Number} copy-timeout A time-out for a state changing, ms.
 */

/** @type {Settings} */
```

Reply:
Yes, JSDoc has a history of being slow to adopt TypeScript-style syntax and common TypeScript types (e.g. `Record`).

I personally don't worry about writing JSDoc-compatible JSDoc annotation as long as TS can parse them. JSDoc is just a static site generator and isn't even used for this file specifically. TypeScript, on the other hand, can use these types to detect mistakes and provide code completion.

Also, JSDoc might be able to parse it if we replace `Record<X, Y>` with `Object<X, Y>`.